### ISSUE TITLE
Object in localns but signature of inner function not resolved

### ISSUE THREAD

**Issue by lucyleeow** [pyapp-kit/in-n-out#93]
### Description

When resolving the signature of a nested inner function, importing a signature object globally works but importing inside the outer function does not (even though this results in the object being in the local namespace of the outer function).

### What I Did

```python
from __future__ import annotations

from in_n_out import type_resolved_signature
from app_model import Action


def outer():
    from magicgui.widgets import FunctionGui
    def myfun(a: FunctionGui):
        if isinstance(a, FunctionGui):
            return a

    print(f'locals {locals()}')

    return Action(id='test', title='title', callback=myfun)

act = outer()
a = type_resolved_signature(act.callback, raise_unresolved_optional_args=False, raise_unresolved_required_args=False)
print(a)
```
gives:

```bash
locals {'myfun': <function outer.<locals>.myfun at 0x7f22eccf6280>, 'FunctionGui': <class 'magicgui.widgets._function_gui.FunctionGui'>}
(a: 'FunctionGui')
```

I have not yet looked into why this is happening but you may know already.

Hopefully I got this right and you would expect this to resolve.

**Comment by tlambert03** [comment]
Actually I would not expect that to resolve.  You've hit on one of the big issues in usage of type annotations at runtime.

This is a long and complicated topic with lots written elsewhere, so I'll just link to relevant peps and you should search them:

- https://lukasz.langa.pl/61df599c-d9d8-4938-868b-36b67fdb4448/ (see the section "where pep 649 helps")
- https://peps.python.org/pep-0563/
- https://peps.python.org/pep-0649/

Long story short, that usage just won't be supported.  You would need to not use `__future__.annotations` or import at the global level

**Comment by tlambert03** [comment]
You can also read through https://github.com/samuelcolvin/pydantic/issues/2678

Point is, what we're trying to do here (using types at runtime) does have some limitations and caveats, and does kinda rely on a little "participation/cooperation" on the part of the end user, just as it does in pydantic

**Comment by lucyleeow** [comment]
Okay interesting, thanks for the info, I'll close this!

**Comment by lucyleeow** [comment]
Thinking about it more, of course it wouldn't work, the outer function is defined but never called until runtime
